File: Bin/Data/LuaScripts/Rotator.lua
-- Rotator script object class. Script objects to be added to a scene node must implement the empty ScriptObject interface
Rotator = ScriptObject()

function Rotator:Start()
    self.rotationSpeed = Vector3(0, 0, 0)
end

function Rotator:SetRotationSpeed(speed)
    self.rotationSpeed = Vector3(speed)
end

function Rotator:Update(timeStep)
    local x = self.rotationSpeed.x * timeStep
    local y = self.rotationSpeed.y * timeStep
    local z = self.rotationSpeed.z * timeStep
    self.node:RotateXYZ(x, y, z)
end

File: Source/Samples/10_RenderToTexture/Rotator.cpp
// Copyright (c) 2008-2022 the Urho3D project
// License: MIT

#include <Urho3D/Scene/Scene.h>

#include "Rotator.h"

#include <Urho3D/DebugNew.h>

Rotator::Rotator(Context* context) :
    LogicComponent(context),
    rotationSpeed_(Vector3::ZERO)
{
    // Only the scene update event is needed: unsubscribe from the rest for optimization
    SetUpdateEventMask(USE_UPDATE);
}

void Rotator::SetRotationSpeed(const Vector3& speed)
{
    rotationSpeed_ = speed;
}

void Rotator::Update(float timeStep)
{
    // Components have their scene node as a member variable for convenient access. Rotate the scene node now: construct a
    // rotation quaternion from Euler angles, scale rotation speed with the scene update time step
    node_->Rotate(Quaternion(rotationSpeed_.x_ * timeStep, rotationSpeed_.y_ * timeStep, rotationSpeed_.z_ * timeStep));
}
